Support predefined jwt section as additional scope source (with additional cosmetics)

michaelklishin opened the pull request:
This is #40 by @papugamichal with various cosmetic changes. See #40 for background.

Commit:
Implement support for gathering scopes from predefined JWT section and combine them with existing ones in post_process_payload () method. Create unit_SUITE and system_SUITE test cases.

Commit:
Resolve PR comments - rename variables.

michaelklishin committed:
Make this code less unorthodox, take 1

michaelklishin committed:
Make this code less unorthodox, take 2

Also improves naming a bit.

acogoluegnes commented on src/rabbit_auth_backend_oauth2.erl:
```diff
@@ -145,37 +150,75 @@ check_token(Token) ->
+            TokenScopes = maps:get(<<"scope">>, Payload),
+            case TokenScopes of
+                [] -> maps:put(<<"scope">>, AdditionalScopes, Payload);
+                _  -> maps:put(<<"scope">>, AdditionalScopes ++ TokenScopes, Payload)
```
`scope` can still be [a binary or an array](#24) here, they're transformed in the last statement of `post_process_payload`. I'll add a test to see if this case is handled (complex claim is on and scope is a binary).

papugamichal replied:
Fair point, i overlooked this case. Thank you!

acogoluegnes replied:
I ended moving up the binary to array conversion, so complex scope and keycloak will always see an array downstream.

acogoluegnes commented on src/rabbit_auth_backend_oauth2.erl:
```diff
@@ -145,37 +150,75 @@ check_token(Token) ->
-            maps:put(<<"scope">>, Scopes, Payload)
+            TokenScopes = maps:get(<<"scope">>, Payload),
+            case TokenScopes of
+                [] -> maps:put(<<"scope">>, AdditionalScopes, Payload);
```
Same as above. Maybe the simplest would be to transform `scope` into an array at the beginning of `post_process_payload` and we should be fine afterwards.

acogoluegnes replied:
BTW, @papugamichal why do we change the Keycloak part? The initial request was `permissions` would replace `scope` alltogether and Keycloak did not complain about this implementation. What's the rationale here?

@michaelklishin WDYT?

papugamichal replied:
@acogoluegnes previous implementation replaces scopes which came with JWT. Was it desirable behaviour?

papugamichal replied:
This is edge case - if token contains both ComplexClaims and Keycloack's scopes then post_process_payload_keycloak () will override existing ones (present in JWT) and those comming from post_process_payload_complex_claim(). IMO, this should be changed (see above implementation) or documentation should be updated to information about "side effects".

acogoluegnes replied:
Yes, the previous implementation would replace the scope alltogether. But I see your point: the complex scope processing could add some scopes that would then be erased by the keycloak processing. This should not happen the way you implemented it, this is defensive for an edge case, so we can keep it.

acogoluegnes reviewed (commented):
This looks OK overall, but needs a bit of rework. I'll polish the tests, add some, and check the system tests (not sure the environment changes are actually applied).

acogoluegnes committed:
Polish extra scopes source tests

Set up environment in init/end test functions, change some scopes in
test to make assertion more obvious.

References #41

acogoluegnes committed:
Move up aud/scope binary to array conversion

This way scope is always an array which makes it easier for downstream
processing.

References #41

michaelklishin commented:
Backported to `v3.8.x`.

papugamichal commented:
@michaelklishin @acogoluegnes  thank you for review and support with handling this PR.
